### PR TITLE
EOS-24914: SSL certificate: Generate ssl certificate if its not there in the specified path

### PR DESCRIPTION
**Pull request by rohitkolapkar**
# Backend

# Problem Statement
[EOS-24914](https://jts.seagate.com/browse/EOS-24914)
[EOS-25351](https://jts.seagate.com/browse/EOS-25351)

- _Overview: JIRA number/GitHub Issue added to PR: [Y/N]:_ Y
- _Describe the problem this patch intends to solve._
If csm_agent protocol is 'https' and ssl_certificate is not present at specified file path then, this patch will call to ssl_generation utility from cortx py-utils that will create a self-signed ssl certificate at specified location with given dns name list and validity. 

## Design

- _For Bug describe the fix here._
- _For feature, post the link to the solution page._
- _Any northbound interface change and has been notified to other gatekeepers [Y/N]:_

## Coding

- _Coding conventions are followed and code is consistent. Yes/No?_ Yes
- _Confirm All CODACY errors are resolved. Yes/No?_ Yes

## Testing

- _Confirm that Test Cases are added (for both the cases, fix and feature). Yes/No?_
- _Confirm Test Cases cover Happy Path, Non-Happy Path and Scalability. Yes/No?_
- _Confirm Testing was performed with installed RPM. Yes/No?_
- _Confirm all the ut/regression/smoke test has been performed [Y/N]_

1. Tested Mini provisioning Phases: PASS
image

2. Tested newly generated ssl_certificate for running csm_agent: PASS
image
image

## Checklist

- _PR is self-reviewed? Yes/No?_ Yes
- _GitHub Issue is updated_ yes
- _Jira is updated_ yes
  -  _Check if the description is clear and explained._  yes
    -  _Check Acceptance Criterion is defined._ yes
      -  _All the tests performed should be mentioned before Resolving a JIRA._ 
        -  _Verification needs to be done before marked as Closed/Verified_
        - _Any interface change Yes/No?_
        - _Change notified to other gatekeepers: Yes/No?_
        - _Code reviews done and addressed: Yes/No?_
        - _Codacy issues reviewed and addressed: Yes/No?_ yes
        - _Deployment test : Done/Not?_
        - _UT/smoke test: Done/Not?_
        - _Jira number added to PR: Yes/No?_ yes
        - _Github merge done using UI: Yes/No?_
        - _Side effects on other features - deployment/update/node-replacement_
        - _Changes to quick-start-guide/community impact_
        - _All dependent component code PR are raised or already merged - **PR is raised. Not yet merged**
        https://github.com/Seagate/cortx-utils/pull/578
        - _All dependent code already merged in landing branch -  **No**

### CODE REVIEW

File: csm/conf/post_install.py [hunk 1]
```diff
@@ -127,10 +128,26 @@ def validate_3rd_party_pkgs(self):
             raise CsmSetupError(f"Failed at package Validation: {ve}")
 
     def set_ssl_certificate(self):
-        ssl_certificate = Conf.get(const.CONSUMER_INDEX, self.conf_store_keys[const.KEY_SSL_CERTIFICATE])
-        Conf.set(const.CSM_GLOBAL_INDEX, const.SSL_CERTIFICATE_PATH, ssl_certificate)
-        Conf.set(const.CSM_GLOBAL_INDEX, const.PRIVATE_KEY_PATH_CONF, ssl_certificate)
-        Log.info(f"Setting ssl certificate path: {ssl_certificate}")
+        ssl_certificate_path = Conf.get(const.CONSUMER_INDEX, self.conf_store_keys[const.KEY_SSL_CERTIFICATE])
```
Comment (kupranay) [inline csm/conf/post_install.py:131]: ssl_certificate_path should be extracted only when its required. Move it inside if csm_protocol== "https" block

File: csm/conf/post_install.py [hunk 2]
```diff
@@ -127,10 +128,26 @@ def validate_3rd_party_pkgs(self):
...
+        csm_endpoint = Conf.get(const.CONSUMER_INDEX, const.CSM_AGENT_ENDPOINTS_KEY)
+        csm_protocol, csm_host, csm_port = self._parse_endpoints(csm_endpoint)
```
Comment (kupranay) [inline csm/conf/post_install.py:133]: i think csm_endpoint is being used only once here in full code. if my assumption is correct then adding a variable here, we are unnecessary adding work for garbage collector. We should avoid it

File: csm/conf/post_install.py [hunk 3]
```diff
@@ -127,10 +128,26 @@ def validate_3rd_party_pkgs(self):
...
+        if csm_protocol == 'https' and not os.path.exists(ssl_certificate_path):
+            Log.warn(f"HTTPS enabled but SSL certificate not found at: {ssl_certificate_path}")
+            Log.info(f"Generating self signed ssl certificate")
```
Comment (kupranay) [inline csm/conf/post_install.py:136]: "Generating self signed ssl certificate" should be combined with log.warn

File: csm/conf/post_install.py [hunk 4]
```diff
@@ -127,10 +128,26 @@ def validate_3rd_party_pkgs(self):
...
+            try:
+                ssl_cert_configs = {"country" : "IN", "state" : "MH", "locality" : "Pune",
```
Comment (kupranay) [inline csm/conf/post_install.py:138]: lets move ssl_cert config to constants

**Review by kupranay — CHANGES_REQUESTED** [review]

File: csm/conf/post_install.py [hunk 5]
```diff
@@ -127,10 +128,26 @@ def validate_3rd_party_pkgs(self):
...
+                                    "organization" : "Seagate Technology", "CN" : "seagate.com"}
+                ssl_cert = Certificate.factory('ssl')
+                ssl_cert_obj = ssl_cert(cert_path = ssl_certificate_path, dns_list = const.DNS_LIST,
+                                            **ssl_cert_configs)
+                ssl_cert_obj.generate()
+            except Exception as e:
```
Comment (pawankumarsrivastava) [inline csm/conf/post_install.py:144]: You are raising SSLCertificateError from utils, so you should catch SSLCertificateError not the base exception

File: csm/conf/post_install.py (hunk 3 above)
Comment (rohitkolapkar) [reply csm/conf/post_install.py]: done

File: csm/conf/post_install.py (hunk 4 above)
Comment (rohitkolapkar) [reply csm/conf/post_install.py]: done

File: csm/conf/post_install.py [hunk 6]
```diff
@@ -127,10 +128,26 @@ def validate_3rd_party_pkgs(self):
...
+        ssl_certificate_path = Conf.get(const.CONSUMER_INDEX, self.conf_store_keys[const.KEY_SSL_CERTIFICATE])
+        csm_endpoint = Conf.get(const.CONSUMER_INDEX, const.CSM_AGENT_ENDPOINTS_KEY)
+        csm_protocol, csm_host, csm_port = self._parse_endpoints(csm_endpoint)
+        if csm_protocol == 'https' and not os.path.exists(ssl_certificate_path):
+            Log.warn(f"HTTPS enabled but SSL certificate not found at: {ssl_certificate_path}")
+            Log.info(f"Generating self signed ssl certificate")
+            try:
+                ssl_cert_configs = {"country" : "IN", "state" : "MH", "locality" : "Pune",
+                                    "organization" : "Seagate Technology", "CN" : "seagate.com"}
+                ssl_cert = Certificate.factory('ssl')
```
[the remark this answers is not in the document]
Comment (rohitkolapkar) [reply csm/conf/post_install.py]: done

File: csm/conf/post_install.py (hunk 5 above)
Comment (rohitkolapkar) [reply csm/conf/post_install.py]: done

**Comment by rohitkolapkar** [comment]
This PR is dependent on utils PR - https://github.com/Seagate/cortx-utils/pull/578

@kupranay  @pawankumarsrivastava

**Review by kupranay — APPROVED** [review]

**Review by pawankumarsrivastava — APPROVED** [review]

**Pull request merged**